### ISSUE TITLE
Method validateOrder automatically send order confirmation email

### ISSUE THREAD

**Issue by viglu** [PrestaShop/PrestaShop#21995]
#### Updated issue content

After reviewing my payment gateway ; it is easier to generate the order reference by my own and then passing it to validateOrder.

Therefore my PR #24357.

The Workflow for the customer with my payment gateway Digicash (https://github.com/viglu/paymentdigicash) is
1. Customer select payment Digicash
2. Customer is redirected to a dedicated payment page. In the background the order ref is generate and display in clear text but also as QR Code to the Customer
3. Customer scans Qr Code with his mobile bank app to pay.
4. A callback from the service provider is made to the paymemt gateway and here validateOrder is called with the generated order ref before and order conf email is sent out.
5. Customer is then redirected to the order confirmation page.

#### Initial issue content (evolved in time)

I'm developing a payment module. I need to submit the Order Reference like SUVOYMMFK. For that I can call validateOrder to create my order. The problem with validateOrder is that is sends automatically the order confirmation email.

#### Describe the solution you'd like (evolved in time)

I sugguest to add a protected variable sendOrderConfirmationEnabled by default = true
This variable sendOrderConfirmationEnabled can be set in my module to false
Put the content of of the IF statement after the comment "// Send an e-mail to customer (one order = one email)" in a seperate function like sendOrderConfirmation($orderId)
Extend the condition of this IF with sendOrderConfirmationEnabled.
After that is possible to call validateOrder without sending the order confirmation, but it could be send seperatly by calling the new function sendOrderConfirmation  after the payment is really done.
And this only to get the Order Reference which need to be delivered to the service provider.

**Comment by Arman-Hosseini** [comment]
Hey @viglu, It sounds good!

If it's very necessary, you can **override valideOrder** method as a quick way and put all the code without send email section, otherwise i can work on it next week.

**Comment by hibatallahAouadni** [comment]
Hello @viglu 

Thanks for the suggestion, the Product Team will take a look.
Please stay tuned !

Ping @PrestaShop/product-team , what do you think?

**Comment by hibatallahAouadni** [comment]
Ping @PrestaShop/prestashop-core-developers wdyt?

**Comment by viglu** [comment]
An other approach is to add another parameter to validateOrder, like $order_ref = null.
And in validateOrder do something like

```php
if ($order_ref == null) {
            do {
                $reference = Order::generateReference();
            } while (Order::getByReference($reference)->count());
else {
            $reference = $order_ref; 
}
```

With this solution nearly no work has do be done.
For me side I could generate the order_ref before and pass it to the payment service provider and the order should be save in DB with my order_ref.

**Comment by MatShir** [comment]
Hi @viglu, is the validateOrder triggered by a type of status or when the merchant's client finishes the checkout?

**Comment by viglu** [comment]
validateOrder is triggered when client has finished the payment, when the service prodiver says: payment ok, after that customer is redirected to order confirmation

**Comment by PierreRambaud** [comment]
It seems legit to be able to generate your own order reference then use the validate order method without sending the mail :+1:

**Comment by MatShir** [comment]
Agreed with @PierreRambaud, for the online payment module, the confirmation email can wait for the validation of the payment, which avoids sending the confirmation email if the payment fails.

**Comment by MatShir** [comment]
Hi @viglu,
Your feature is shared with the team. But in the meantime, PrestaShop is an open-source project, if you have the technical knowledge or have someone that could don't hesitate to propose a pull request ([find more information here](https://devdocs.prestashop.com/1.7/contribute/contribute-pull-requests/)). We will be happy to add your feature to the PrestaShop core.

**Comment by viglu** [comment]
After reviewing my payment gateway it easier to generate the order referenace by my own and then passing it to validateOrder.
Therefore my PR #24357.
The Workflow for the customer with my payment gateway Digicash (https://github.com/viglu/paymentdigicash) is
1. Customer select payment Digicash
2. Customer is redirected to a dedicated payment page. In the background the order ref is generate and display in clear text but also as QR Code to the Customer
3. Customer scans Qr Code with his mobile bank app to pay.
4. A callback from the service provider is made to the paymemt gateway and here validateOrder is called with the generated order ref before and order conf email is sent out.
5. Customer is then redirected to the order confirmation page.

**Comment by matks** [comment]
Thank you @viglu the change will be released with PrestaShop 8.0.0 in 2022
